server/src/main/java/org/opensearch/gateway/BaseShardCache.java

```diff
@@ -0,0 +1,335 @@
+ * Setting up the cache is required from implementation class.
+ * Store node level responses of transport actions like {@link TransportNodesListGatewayStartedShards} or
+ * {@link TransportNodesListShardStoreMetadata} using the given functionalities.
+ * <p>
+ * initData : how to initialize an entry of shard cache for a node.
+ * putData : how to store the response of transport action in the cache.
+ * getData : how to get the stored data for any shard allocators like {@link PrimaryShardAllocator} or
+ * {@link ReplicaShardAllocator}
+ * deleteData : how to clean up the stored data from cache for a shard.
+ *
+ * @param <K> Response type of transport action which has the data to be stored in the cache.
+ */
+public abstract class BaseShardCache<K extends BaseNodeResponse> {
+    private final Logger logger;
+    private final String type;
+
+    protected BaseShardCache(Logger logger, String type) {
+        this.logger = logger;
+        this.type = type;
+    }
+
+    /**
+     * Initialize cache's entry for a node.
+     *
+     * @param node for which node we need to initialize the cache.
+     */
+    abstract void initData(DiscoveryNode node);
+
+    /**
+     * Store the response in the cache from node.
+     *
+     * @param node     node from which we got the response.
+     * @param response shard metadata coming from node.
+     */
+    abstract void putData(DiscoveryNode node, K response);
+
+    /**
+     * Populate the response from cache.
+     *
+     * @param node node for which we need the response.
+     * @return actual response.
+     */
+    abstract K getData(DiscoveryNode node);
+
+    /**
+     * Get actual map object of the cache
+     *
+     * @return map of nodeId and NodeEntry extending BaseNodeEntry
+     */
+    @NonNull
+    abstract Map<String, ? extends BaseShardCache.BaseNodeEntry> getCache();
+
+    /**
+     * Cleanup cached data for this shard once it's started. Cleanup only happens at shard level. Node entries will
+     * automatically be cleaned up once shards are assigned.
+     *
+     * @param shardId for which we need to free up the cached data.
+     */
+    abstract void deleteData(ShardId shardId);
+
+    /**
+     * Returns the number of fetches that are currently ongoing.
+     */
+    int getInflightFetches() {
+        int count = 0;
+        for (BaseNodeEntry nodeEntry : getCache().values()) {
+            if (nodeEntry.isFetching()) {
+                count++;
+            }
+        }
+        return count;
+    }
+
+    /**
+     * Fills the shard fetched data with new (data) nodes and a fresh NodeEntry, and removes from
+     * it nodes that are no longer part of the state.
+     */
+    void fillShardCacheWithDataNodes(DiscoveryNodes nodes) {
+        // verify that all current data nodes are there
+        for (final DiscoveryNode node : nodes.getDataNodes().values()) {
+            if (getCache().containsKey(node.getId()) == false) {
+                initData(node);
```

Comment:
isn't this BaseShardCache just suppose to be abstraction for HashMap for cache so that you have more memory efficient implementation, why do you need all these methods here (including the below ones)? how do they belong to cache?

Reply:
Hi @shwetathareja, 

I created one interface in previous commit which had clear 4 functionalities that need to be implemented by actual cache implementations. https://github.com/opensearch-project/OpenSearch/pull/12441#discussion_r1513181561

But I removed it and moved back the methods in this abstract class. As @sachinpkale  suggested, having a separate interface whose type we're not using anywhere is complicated for user. So, better to keep these methods in abstract class itself. 
Please suggest if you think otherwise.

### Why I moved the methods from AsyncShardFetch class to BaseShardCache 
As we know these caches are only going to store the data at node level because they store the transport responses of broadcast actions. Common functionalities like checking if fetching is done or not or initializing node entry etc all should be in one place so `AsyncShardFetch` can just use **cache** parameter to call all these methods, else `AsyncShardFetch` will need to get the cache and perform these operations itself. 

So now the responsibilities of classes are pretty clear with this move. 

AsyncShardFetch - Responsible for doing actual asyncFetch (transport call) and use the cache object to init, process and get the cache data.
BaseShardCache - Abstract class keeping all necessary common functionalities of cache which is needed by both `ShardCache` or `ShardBatchCache`. 
ShardCache - Current implementation of per shard cache (normal hashmap)
[ShardBatchCache](https://github.com/opensearch-project/OpenSearch/pull/12504/files#diff-9f14f4320d39587e9cb6d8a76bed25ed1a2ee8464246700ccb3a3777b4ccb516) - new memory efficient implementation for a batch of shards (specifically for new written batch transport actions)

Reply:
In latest commit, I've moved the ShardCache class as inner class of AsyncShardFetch to keep is restricted to that class only. And changed the name of BaseShardCache to `AsyncShardFetchCache` to make it more specific of this use case.